docs/library/network.rst

```diff
@@ -29,6 +29,95 @@ For example::
     data = s.recv(1000)
     s.close()
 
+Common network interface
+========================
+
+.. class:: BaseNIC(id=None, ...)
+
+Create a network interface object. Parameters are network interface
+dependent. If there are few interfaces of the same type, the first
+parameter should be `id`.
+
+    .. method:: active([is_active])
+
+        Activate ("up") or deactivate ("down") network interface, if boolean
+        argument is passed. Otherwise, query current state if no argument is
+        provided. Most other methods require an active interface.
+
+    .. method:: connect([service_id, key=None, \*, ...])
+
+       Connect interface to the network. This method is optional, and
+       available only for interfaces which are not "always connected".
+       If no parameters are given, connect to the default (or only)
+       service. If a single parameter is given, it is primary identifier
+       of a service to connect to. It may be accompanied by a key
+       (password) required to access service. There can be further
+       arbitrary keyword-only parameters, depending on networking medium
+       type and/or particular device. Parameters can be used to: a)
+       specify alternative service identifer types; b) provide additional
+       connection parameters. For various medium types, there are different
+       sets of predefined/recommended paramaters, among them:
+
+       * WiFi: `bssid` keyword to connect by BSSID (MAC address) instead
+         of access point name
+
+    .. method:: disconnect()
+
+       Disconnect from network.
+
+    .. method:: isconnected()
+
+       Returns True if connected to network.
+
+    .. method:: scan(\*, ...)
+
+       Scan for the available network services/connections. Returns
+       list of tuples with discovered service parameters. For various
+       network media, there are different variants of predefined/
+       recommended tuple formats, among them:
+
+       * WiFi: (ssid, bssid, channel, RSSI, authmode, hidden). There
+       may be further fields, specific to a particular device.
+
+       The may accept additional keyword arguments to filter scan results
+       (e.g. scan for a particular service, on a particular channel,
+       for services of a particular set, etc.), and to affect scan
+       duration and other parameters. Where possible, parameter names
+       should match those in connect().
+
+    .. method:: status()
+
+       Return detailed status of the interface, values are dependent
+       on the network medium/technology.
+
+    .. method:: ifconfig([(ip, subnet, gateway, dns)])
```

Comment:
So, work is being done here is two-fold:

* Find a "union" of the functionality already implemented in existing network modules.
* Define a "superset" of that functionality, as much as possible in a compatible way.

So, "None as placeholders in the tuple for elements that you don't want to change" - is it implemented for any of the mainline platforms? If not, what's the point of documenting it if doesn't work?

Beyond that, what if a user wants to "undefine" gateway or DNS? "None" is the natural value for that, yet you propose to assign it a special meaning.

This shows that designing APIs a tricky matter. While I empathize with the idea to make it "easier", I'm afraid using your 2-liner is the best option we have - now and for the foreseeable future. Again, the idea is first to try to capture what we already have implemented, e.g. things which were implemented for esp8266, but not documented. Then, things which users want to be implemented for esp8266 (some of these things contradict one another!).